extras/devices/LVFS/LVFS.py

```diff
@@ -0,0 +1,312 @@
+
+        :param fwupd_raw: output string of ```$ fwupdmgr get-devices```
+        """
+        for dev in json.loads(fwupd_raw)["Devices"]:
+            if "Flags" in dev and "updatable" in dev["Flags"]:
+                self.fw_info.append(dev)
+
+        for dev in self.fw_info:
+            if "Version" in dev:
+                dev_name = dev["Name"] if "Name" in dev else dev["DeviceId"]
+                msg = f"[{dev_name}] current version: {dev['Version']}"
+            else:
+                msg = ""
+            if "Releases" in dev:
+                higher_ver, lower_ver, same_ver = [], [], []
+                vers = ""
+                for rel in dev["Releases"]:
+                    if "Flags" not in rel:
+                        continue
+                    elif "is-upgrade" in rel["Flags"]:
+                        higher_ver.append(rel["Version"])
+                    elif "is-downgrade" in rel["Flags"]:
+                        lower_ver.append(rel["Version"])
+                    else:
+                        same_ver.append(rel["Version"])
+                if higher_ver != []:
+                    vers = f" - LVFS upgradable version(s): {', '.join(higher_ver)}"
+                if lower_ver != []:
+                    vers += f" - LVFS downgradable version(s): {', '.join(lower_ver)}"
+                if vers:
+                    msg += vers
+                    print(msg)
+            else:
+                msg += " - no available firmware on LVFS"
+            if msg:
+                logger.info(msg)
+
+    def upgrade(self) -> bool:
+        """
+        Upgrade all devices firmware to latest version if available on LVFS
+
+        :return: `True` if upgrading is done and reboot is required, `False`
+                 otherwise
+        """
+        logger.info("start upgrading")
+        reboot = False
+        for dev in self.fw_info:
+            dev_name = dev["Name"] if "Name" in dev else dev["DeviceId"]
+            try:
+                latest_ver = dev["Releases"][0]
+            except KeyError:
+                continue
+            if dev["Version"] != latest_ver["Version"]:
+                if "is-upgrade" in latest_ver["Flags"]:
+                    logger.info(
+                        "[%s] try upgrading to %s"
+                        % (dev_name, latest_ver["Version"])
+                    )
+                    dev["targetVersion"] = latest_ver["Version"]
+                    rc, stdout, stderr = self.run_cmd(
+                        f"sudo fwupdmgr upgrade {dev['DeviceId']} -y --no-reboot-check",
+                        raise_stderr=False,
+                    )
+                    if rc == 0:
+                        logger.debug(stdout)
+                        reboot = True
+                    else:
+                        logger.error(
+                            "[%s] Failed to upgrade %s"
+                            % (dev_name, latest_ver["Version"])
+                        )
```

Comment:
logging works a bit differently. Instead of using %s usually you specify it the following way
```suggestion
                        logger.error(
                            "[%s] Failed to upgrade %s"
                            dev_name, 
                            latest_ver["Version"],
                        )
```

Reply:
You can see such usage in the Python docs: https://docs.python.org/3/howto/logging.html#optimization
```python
    logger.debug('Message with %s, %s', expensive_func1(),
                                        expensive_func2())
```

Reply:
Thank you for the detailed explanation!